p2p/dial_test.go

```diff
@@ -0,0 +1,325 @@
+package p2p
+
+import (
+	"encoding/binary"
+	"reflect"
+	"testing"
+	"time"
+
+	"github.com/davecgh/go-spew/spew"
+	"github.com/ethereum/go-ethereum/p2p/discover"
+)
+
+func init() {
+	spew.Config.Indent = "\t"
+}
+
+type dialtest struct {
+	init   *dialstate // state before and after the test.
+	rounds []round
+}
+
+type round struct {
+	peers []*Peer // current peer set
+	done  []task  // tasks that got done this round
+
+	new []task // the result must match this one
+}
+
+func runDialTest(t *testing.T, test dialtest) {
+	var (
+		vtime   time.Time
+		running []task
+	)
+	pm := func(ps []*Peer) map[discover.NodeID]*Peer {
+		m := make(map[discover.NodeID]*Peer)
+		for _, p := range ps {
+			m[p.rw.id] = p
```

Comment:
Maybe we could add an existence check here? It's just a simple sanity check to prevent duplicate peers. Currently if I duplicate anything in the round.peers, all the tests will still happily pass. If you deem it not worthy then we can skip it, no problem. The could would be something like:

```
			id := p.rw.id
			if _, exists := m[id]; exists {
				t.Errorf("duplicate peer: %v", id)
			} else {
				m[id] = p
			}
```